chore: Rework how artifact download is handled

vpodzime opened the pull request:
In order to avoid passing many callbacks around as arguments and calling the update module's callback 9 levels of nesting deep, we just call the particular functions directly wherever possible. This requires a separate code path tailored to artifact download.

Ticket: MEN-7499
Changelog: none

vpodzime commented:
This is what I was able to put together for https://northerntech.atlassian.net/browse/MEN-7499. However, it is here only for preview/discussion/suggestions purposes because it's quite wrong in at least two things:
- code duplication
- twisted dependencies between code units, e.g. `platform/net/zephyr/src/mender-http.c` having `#include "mender-api.h"` or `core/src/mender-artifact.c` having `#include "mender-client.h"`

And because I believe the above two issues are bad, especially the second one, I didn't bother adding the `posix` platform implementation of the `mender_http_artifact_download()` function, even though it would be easy because on POSIX it could just use the same callback passing as it does on the `main` branch.

Last but not least, this doesn't seem to fix the mysterious hangs and crashes. (although I keep tracking those down the :rabbit: :hole: !)

Feedback welcome!

vpodzime commented:
I hope to push this into a reasonable state during today and tomorrow (working from :car:)

larsewi reviewed (approved):
Great work 🚀  I know you mostly been moving code around, so feel free to dismiss my suggestions.

lluiscampos commented on core/src/mender-artifact-download.c:
```diff
@@ -0,0 +1,120 @@
+ * @file      mender-artifact-download.c
+ * @brief     Mender artifact download implementation
+ *
+ * Copyright Northern.tech AS
```
Same question as for the other file.

vpodzime replied:
This one has Joel's code, that's true. Adding.

lluiscampos commented on core/src/mender-artifact.c:
```diff
@@ -1119,15 +1187,16 @@ artifact_read_data(mender_artifact_ctx_t *ctx, mender_err_t (*callback)(char *,
-                               ctx->file.index,
-                               length))) {
+        /* Invoke the download artifact callback */
+        ret = process_artifact_data_callback(ctx->payloads.values[index].type,
```
From this line... such a simple change :smile:

vpodzime replied:
?

lluiscampos replied:
It was meant a compliment. If you only look at this line your PR seems "so simple", yet the work is everywhere else to make this possible :rocket:

lluiscampos commented on platform/net/zephyr/src/mender-http.c:
```diff
@@ -58,6 +60,14 @@ static mender_http_config_t mender_http_config;
+ * @param final_call Indicate final call
+ * @param user_data User data, used to retrieve request context data
+ */
+static void artifact_response_cb(struct http_response *response, enum http_final_call final_call, void *user_data);
```
For consistency with the above cb: `mender_artifact_response_cb`

vpodzime replied:
Actually, I think it should be the other way around -- no `mender_` prefix for `static` functions.

lluiscampos commented on platform/net/zephyr/src/mender-http.c:
```diff
@@ -257,6 +371,26 @@ mender_http_response_cb(struct http_response *response, enum http_final_call fin
 }
 
+static void
+artifact_response_cb(struct http_response *response, MENDER_ARG_UNUSED enum http_final_call final_call, void *user_data) {
```
Pre-existing,

Please add `MENDER_ARG_UNUSED enum...` to `mender_http_response_cb` also.

lluiscampos commented on platform/net/zephyr/src/mender-http.c:
```diff
@@ -257,6 +371,26 @@ mender_http_response_cb(struct http_response *response, enum http_final_call fin
+    assert(NULL != user_data);
+
+    /* Retrieve request context */
+    mender_artifact_download_data_t *dl_data = user_data;
```
We should cast the void pointer to `(mender_artifact_download_data_t *) user_data`

vpodzime replied:
That's done by the assignment.

lluiscampos replied:
TIL! Thanks

lluiscampos reviewed (approved):
Great work :rocket:  Really nice commit message, thank you.

vpodzime commented on core/src/mender-artifact-download.c:
```diff
@@ -0,0 +1,120 @@
+                                                      mender_artifact_download_data_t *dl_data);
+
+mender_err_t
+mender_download_artifact(char *uri, mender_deployment_data_t *deployment_data, mender_update_module_t **update_module) {
```
Yup, exactly my thinking. But then I saw I'd have to change the functions this is passed into... Will address this in a fw-up.

vpodzime commented on include/mender-http-client-event.h:
```diff
@@ -0,0 +1,42 @@
+ * @file      mender-http-client-event.h
+ * @brief     Mender HTTP client event type
+ *
+ * Copyright joelguittet and mender-mcu-client contributors
```
Yes, as long as they contain Joel's code.

vpodzime commented on core/src/mender-update-module.c:
```diff
@@ -0,0 +1,77 @@
+ * @file      mender-update-module.c
+ * @brief     Mender update Module implementation
+ *
+ * Copyright Northern.tech AS
```
Yup! Good catch!

vpodzime replied:
Actually, not really. I wrote this code for update modules, right?

vpodzime commented:
@larsewi @lluiscampos I addressed the issues introduced in this PR and I'm merging it as-is now. Will follow up with a PR addressing the unresolved pre-existing issues you pointed out.

vpodzime commented on include/mender-client.h:
```diff
@@ -137,6 +129,21 @@ mender_err_t mender_client_ensure_authenticated(void);
+ * @param length Artifact data length
+ * @return MENDER_OK if the function succeeds, error code if an error occurred
+ */
+mender_err_t mender_client_download_artifact_callback(char *type, cJSON *meta_data, char *filename, size_t size, void *data, size_t index, size_t length);
```
This should no longer exist, actually.

vpodzime commented on core/src/mender-update-module.c:
```diff
@@ -0,0 +1,77 @@
+        for (size_t update_module_index = 0; (NULL == ret) && (update_module_index < update_modules_count); update_module_index++) {
+            /* Check artifact type */
+            if (StringEqual(artifact_type, update_modules_list[update_module_index]->artifact_type)) {
+                ret = update_modules_list[update_module_index];
```
> Shouldn't we `break` here?

The loop condition checks `ret` and stops once it's set.

> Or amend the registering function?

Yes, I think we should not allow multiple UMs for the same artifact type.
